### PR TITLE
Removed redundant copy of Tensor in TF Const translator

### PR DESCRIPTION
**Pull request by popovaan**
Description: Removed redundant copy of Tensor in TF Const translator

Ticket: 113207

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names
* [x]  Transformation preserves tensor names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests
* [x]  Transformation tests
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list
* [x]  Guide on how to convert the **public** model
* [x]  User guide update

### CODE REVIEW

**Review by rkazants — APPROVED** [review]

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py [hunk 1]
```diff
@@ -51,10 +51,26 @@ def __init__(self, operation: tf.Operation, inner_graph: bool):
...
+
+            if value.tensor_content:
+                shape = [d.size for d in value.tensor_shape.dim]
+                tensor_dtype = tf.dtypes.as_dtype(value.dtype)
+                dtype = tensor_dtype.as_numpy_dtype
+                # no copy of content
+                self.m_parsed_content = (np.frombuffer(value.tensor_content,
+                          dtype=dtype).reshape(shape))
+            else:
+                # TODO: remove copy of content for cases when tensor value is not in tensor_content field
```
Comment (rkazants) [inline src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:72]: please list a JIRA ticket for this

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py [hunk 2]
```diff
@@ -118,7 +134,7 @@ def get_attribute(self, name):
                 if isinstance(self.m_parsed_content, np.ndarray):
                     return OVAny(Tensor(self.m_parsed_content))
                 return OVAny(Tensor(np.array([self.m_parsed_content]), shape=[1]))
-            ov_tensor = Tensor(self.m_parsed_content, shared_memory=True)
+            ov_tensor = Tensor(self.m_parsed_content, shared_memory=self.m_shared_memory)
             ov_tensor = OVAny(ov_tensor)
```
Comment (rkazants) [inline src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:138]: is it fine that we always wrap this into `ov::Any`? In case unwrapping, don't we have extra copy of `ov::Tensor`?

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py [hunk 3]
```diff
@@ -51,10 +51,26 @@ def __init__(self, operation: tf.Operation, inner_graph: bool):
                                                     "Expected tf.Operation, got {}".format(type(operation))
         self.m_operation = operation
         self.m_inner_graph = inner_graph
+        self.m_shared_memory = True
         if self.m_operation.type == "Const":
+            # Copies tensor value to parsed TensorProto
             value = self.m_operation.node_def.attr["value"].tensor
-            # copies tensor value from node_def
-            self.m_parsed_content = tf.make_ndarray(value)
+
```
Comment (rkazants) [inline src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:58]: Also, I think before sharing constant/variable values, we need to provide ovc/mo option for switching on/off sharing?
Besides, what mechanism cares that `tf.Variable` will stay in the memory while `ov::Model` is alive

**Review by rkazants — CHANGES_REQUESTED** [review]
The main topic is changed to sharing variable. Now it is not just a fix

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py (hunk 1 above)
Comment (popovaan) [reply src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:72]: Done: CVS-114797

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py (hunk 2 above)
Comment (popovaan) [reply src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:138]: Yes, I checked this part, no copy is done by ov::Any(): 
https://github.com/openvinotoolkit/openvino/blob/2b795afc09cc5319312ba187d9a117cb4f913538/src/core/src/any.cpp#L76.

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py (hunk 3 above)
Comment (popovaan) [reply src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:58]: Yes, I can provide the option for switching on/off sharing, but the name of the option is not determined yet. So probably we can add it in separate PR when the name is clear and not block this PR?

This mechanism does not cares that `tf.Variable` still alive.
@slyalin mentioned that we can recommend user to keep original model in memory as we do with PyTorch FE. 
Or we can recommend to turn off sharing when the option is ready.

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py [hunk 4]
```diff
@@ -660,6 +661,60 @@ def test_unnamed_saved_model_dir(self, ie_device, precision, ir_version, temp_di
...
+    def test_zero_copy(self, ie_device, precision, ir_version, temp_dir):
+        import tensorflow as tf
+        from openvino.tools.mo import convert_model
+        from openvino.runtime import compile_model
+        class LayerModel(tf.Module):
+            def __init__(self):
+                super(LayerModel, self).__init__()
+                self.var1 = tf.Variable(np.random.rand(3).astype(np.float32), name='var1')
+                self.var2 = tf.Variable(np.random.rand(3).astype(np.float32), name='var2')
+
```
Comment (rkazants) [inline tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py:673]: Should we have another test case with constant?

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py [hunk 5]
```diff
@@ -51,10 +51,26 @@ def __init__(self, operation: tf.Operation, inner_graph: bool):
                                                     "Expected tf.Operation, got {}".format(type(operation))
         self.m_operation = operation
         self.m_inner_graph = inner_graph
+        self.m_shared_memory = True
```
Comment (rkazants) [inline src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:54]: please add `TODO` to bind with MO `sharing` option and create ticket

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py (hunk 1 above)
Comment (popovaan) [reply src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py]: Done.

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py [hunk 6]
```diff
@@ -51,10 +51,26 @@ def __init__(self, operation: tf.Operation, inner_graph: bool):
...
         if self.m_operation.type == "Const":
+            # Copies tensor value to parsed TensorProto
```
[the remark this answers is not in the document]
Comment (popovaan) [reply src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:61]: Actually the copy is performed to empty TensorProto structure from inner buffer of TF_Operation. I changed the comment to make it more clear.

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py [hunk 7]
```diff
@@ -660,6 +661,60 @@ def test_unnamed_saved_model_dir(self, ie_device, precision, ir_version, temp_di
...
+        wf_infer1 = keras_model(test_input).numpy()
+
+        CommonLayerTest().compare_ie_results_with_framework(ov_infer1, {"Identity:0": wf_infer1}, eps)
+
+        # Change value of variables in original model
+        for val in keras_model.variables:
+            arr = val.value().__array__()
+            arr[0] = 0
+            arr[1] = 1
+            arr[2] = 2
```
[the remark this answers is not in the document]
Comment (popovaan) [reply tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py:708]: Done.

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py (hunk 4 above)
Comment (popovaan) [reply tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py:673]: Yes, we can test, that we do not loose constant values from tf.Const. Added such test.

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py [hunk 8]
```diff
@@ -51,13 +51,34 @@ def __init__(self, operation: tf.Operation, inner_graph: bool):
                                                     "Expected tf.Operation, got {}".format(type(operation))
         self.m_operation = operation
         self.m_inner_graph = inner_graph
+
+        # Copies value from inner buffer of TF_Operation to NodeDef class.
+        self.m_node_def = self.m_operation.node_def
+
+        # TODO: Create parameter in convert_model() for turing on/off shared memory, ticket: 114971
```
Comment (rkazants) [inline src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:58]:
```suggestion
        # TODO: Create parameter in convert_model() for turning on/off shared memory, ticket: 114971
```

Comment (rkazants) [reply src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:58]: minor

File: src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py [hunk 9]
```diff
@@ -51,13 +51,34 @@ def __init__(self, operation: tf.Operation, inner_graph: bool):
                                                     "Expected tf.Operation, got {}".format(type(operation))
         self.m_operation = operation
         self.m_inner_graph = inner_graph
+
+        # Copies value from inner buffer of TF_Operation to NodeDef class.
+        self.m_node_def = self.m_operation.node_def
```
Comment (rkazants) [inline src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:56]: why do we need this?

Comment (popovaan) [reply src/bindings/python/src/openvino/frontend/tensorflow/node_decoder.py:56]: When we access `self.m_operation.node_def` TF copies value from inner buffer of `TF_Operation` to `NodeDef` class.
In decoder there are several places when we need to get attributes from `node_def`. So avoid redundant copy of values I save `node_def` in node decoder.

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py [hunk 10]
```diff
@@ -660,6 +661,112 @@ def test_unnamed_saved_model_dir(self, ie_device, precision, ir_version, temp_di
...
 
+    def test_zero_copy(self, ie_device, precision, ir_version, temp_dir):
+        import tensorflow as tf
+        from openvino.tools.mo import convert_model
+        from openvino.runtime import compile_model
+        class LayerModel(tf.Module):
+            def __init__(self):
+                super(LayerModel, self).__init__()
+                self.var1 = tf.Variable([7.5, 5.2, 6.7], name='var1')
+                self.var2 = tf.Variable([5.1, 7.7, 3.9], name='var2')
```
[the remark this answers is not in the document]
Comment (popovaan) [reply tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py]: Done.

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py [hunk 11]
```diff
@@ -660,6 +661,112 @@ def test_unnamed_saved_model_dir(self, ie_device, precision, ir_version, temp_di
...
+                return self.sub_function(input)
+
+        if precision == 'FP32':
+            eps = 1e-4
+        else:
+            eps = 5e-2
+
+        # Create TF model with variables
+        keras_model = LayerModel()
+        test_input = np.array(7.65).astype(np.float32)
```
[the remark this answers is not in the document]
Comment (popovaan) [reply tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py]: Done.

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py [hunk 12]
```diff
@@ -660,6 +661,112 @@ def test_unnamed_saved_model_dir(self, ie_device, precision, ir_version, temp_di
...
+                return input * self.var1 + self.var2
+
+            @tf.function()
+            def __call__(self, input):
+                return self.sub_function(input)
+
+        if precision == 'FP32':
+            eps = 1e-4
+        else:
+            eps = 5e-2
```
[the remark this answers is not in the document]
Comment (popovaan) [reply tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py]: Removed eps here and used np.array_equal.

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py [hunk 13]
```diff
@@ -660,6 +661,112 @@ def test_unnamed_saved_model_dir(self, ie_device, precision, ir_version, temp_di
...
+        # Convert model to OV
+        ov_model = convert_model(keras_model, input_shape=[1])
+        cmp_model = compile_model(ov_model)
+
+        # Check model inference
+        ov_infer1 = cmp_model(test_input, ie_device)
+        fw_infer1 = keras_model(test_input).numpy()
+
+        assert CommonLayerTest().compare_ie_results_with_framework(ov_infer1, {"Identity:0": fw_infer1}, eps)
+        assert CommonLayerTest().compare_ie_results_with_framework(ov_infer1, {"Identity:0": [62.475, 47.48, 55.155003]}, eps)
```
[the remark this answers is not in the document]
Comment (popovaan) [reply tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py]: Replaced with simple values.

File: tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py [hunk 14]
```diff
@@ -660,6 +661,112 @@ def test_unnamed_saved_model_dir(self, ie_device, precision, ir_version, temp_di
...
+
+        feed_dict = {"Input:0": test_input}
+        with tf.compat.v1.Session(graph=tf_graph) as sess:
+            fw_infer1 = sess.run('add:0', feed_dict=feed_dict)
+
+        assert CommonLayerTest().compare_ie_results_with_framework(ov_infer1, {"add:0": fw_infer1}, eps)
+        assert CommonLayerTest().compare_ie_results_with_framework(ov_infer1, {"add:0": [2.6, 9.6, 12.4]}, eps)
+
+        # run Garbage collector
+        gc.collect()
```
[the remark this answers is not in the document]
Comment (popovaan) [reply tests/layer_tests/mo_python_api_tests/test_mo_convert_tf.py:755]: This test checks that we do not loose memory allocated for constants after garbage collection. Added comment.

Regarding `gc`, I don't see such limitation in documentation: https://docs.python.org/3/library/gc.html?highlight=del#gc.collect

And the memory loss can be reproduced with this test on master if we remove copy of tensor here:
https://github.com/openvinotoolkit/openvino/blob/7c1949421f804b53636f5f03877872a327853705/src/frontends/tensorflow_common/src/op/const.cpp#L31

**Review by rkazants — APPROVED** [review]
